gateway/proxyproto/ssmproxy/ssm.go

```diff
@@ -0,0 +1,369 @@
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("failed to unmarshal init packet, reason=%v", err)
+		return
+	}
+
+	if initPacket.TokenValue == nil {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("invalid token, reason=%v", err)
+	}
+
+	// Try parse token
+	tokenConnectionId, err := decodeToken(*initPacket.TokenValue)
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("failed to decode token, reason=%v", err)
+		return
+	}
+
+	if tokenConnectionId != connectionId {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("invalid token, expected=%s, got=%s", connectionId, tokenConnectionId)
+		return
+	}
+
+	log.With("sid", sessionID, "conn", cID).
+		Infof("connection established for connectionId=%s, target=%s, awsClientId=%s, awsClientVersion=%s",
+			connectionId, target, initPacket.ClientId, initPacket.ClientVersion)
+
+	dbConnection, err := models.GetConnectionByTypeAndID(pb.ConnectionTypeSSM.String(), connectionId)
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("failed to get connection by id, reason=%v", err)
+		c.String(http.StatusBadRequest, "Invalid request")
+		return
+	}
+	log.With("sid", sessionID, "conn", cID).
+		Infof("starting websocket connection for connectionId=%s, target=%s, sessionID=%s", connectionId, target, sessionID)
+
+	client, err := grpc.Connect(grpc.ClientConfig{
+		ServerAddress: grpc.LocalhostAddr,
+		Token:         "", // it will use impersonate-auth-key as authentication
+		UserAgent:     "ssm/grpc",
+		Insecure:      appconfig.Get().GatewayUseTLS() == false,
+		TLSCA:         appconfig.Get().GrpcClientTLSCa(),
+		// it should be safe to skip verify here as we are connecting to localhost
+		TLSSkipVerify: true,
+	},
+		grpc.WithOption(grpc.OptionConnectionName, dbConnection.ConnectionName),
+		grpc.WithOption(grpckey.ImpersonateAuthKeyHeaderKey, grpckey.ImpersonateSecretKey),
+		grpc.WithOption(grpckey.ImpersonateUserSubjectHeaderKey, dbConnection.UserSubject),
+		grpc.WithOption("origin", pb.ConnectionOriginClient),
+		grpc.WithOption("verb", pb.ClientVerbConnect),
+		grpc.WithOption("session-id", sessionID),
+	)
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).Errorf("failed connecting to hoop server, reason=%v", err)
+		c.String(http.StatusInternalServerError, "Failed to connect to hoop server")
+	}
+	defer client.Close()
+
+	// Send an open session packet
+	err = client.Send(&pb.Packet{
+		Type: pbagent.SessionOpen,
+		Spec: map[string][]byte{
+			pb.SpecGatewaySessionID:   []byte(sessionID),
+			pb.SpecClientConnectionID: []byte(cID),
+		},
+	})
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("failed sending open session packet to hoop server, reason=%v", err)
+		return
+	}
+
+	// Wait for session open confirmation
+	pkt, err := client.Recv()
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("failed receiving session open confirmation from hoop server, reason=%v", err)
+		return
+	}
```

Comment:
Recv() is not thread safe, so you must consume this in a single go routine. You're calling it again at handleTXPipe.

"	// RecvMsg blocks until it receives a message into m or the stream is
	// done. It returns io.EOF when the stream completes successfully. On
	// any other error, the stream is aborted and the error contains the RPC
	// status.
	//
	// It is safe to have a goroutine calling SendMsg and another goroutine
	// calling RecvMsg on the same stream at the same time, but it is not
	// safe to call RecvMsg on the same stream in different goroutines.
"

There are a few examples on the SSH and Postgres Proxy on how to this